packages/babel-plugin-proposal-destructuring-private/src/util.ts

```diff
@@ -0,0 +1,465 @@
+      }
+    }).next().done === false
+  );
+}
+
+export function hasPrivateClassElement(node: t.ClassBody): boolean {
+  return node.body.some(element =>
+    isPrivateName(
+      // @ts-expect-error: for those class element without `key`, they must
+      // not be a private element
+      element.key,
+    ),
+  );
+}
+
+/**
+ * Traverse the given pattern and report the private key path.
+ * A private key path is analagous to an array of `key` from the pattern NodePath
+ * to the private key NodePath. See also test/util.skip-bundled.js for an example output
+ *
+ * @export
+ * @param {t.LVal} pattern
+ */
+export function* privateKeyPathIterator(pattern: t.LVal) {
+  const indexPath = [];
+  yield* traversePattern(pattern, function* (node, index, depth) {
+    indexPath[depth] = index;
+    if (isObjectProperty(node)) {
+      const propertyKey = node.key;
+      if (isPrivateName(propertyKey)) {
+        yield indexPath.slice(1, depth + 1);
+      }
+    }
+  });
+}
+
+type LHS = Exclude<t.LVal, t.RestElement | t.TSParameterProperty>;
+
+type ExcludingKey = {
+  key: t.ObjectProperty["key"];
+  computed: t.ObjectProperty["computed"];
+};
+type Item = {
+  left: LHS;
+  right: t.Expression;
+  restExcludingKeys?: ExcludingKey[] | null;
+};
+
+function rightWillBeReferencedOnce(left: LHS) {
+  switch (left.type) {
+    // Skip memoising the right when left is an identifier or
+    // an array pattern
+    case "Identifier":
+    case "ArrayPattern":
+      return true;
+    case "ObjectPattern":
+      return left.properties.length === 1;
+    default:
+      return false;
+  }
+}
+/**
+ * Transform private destructuring. It returns a generator
+ * which yields a pair of transformed LHS and RHS, which can form VariableDeclaration or
+ * AssignmentExpression later.
+ *
+ * @export
+ * @param {LHS} left The root pattern
+ * @param {t.Expression} right The initializer or the RHS of pattern
+ * @param {Scope} scope The scope where memoized id should be registered
+ * @param {boolean} isAssignment Whether we are transforming from an AssignmengExpression of VariableDeclaration
+ * @returns {Generator<Transformed, void, void>}
+ */
+export function* transformPrivateKeyDestructuring(
+  left: LHS,
+  right: t.Expression,
+  scope: Scope,
+  isAssignment: boolean,
+  shouldPreserveCompletion: boolean,
+  addHelper: File["addHelper"],
+  objectRestNoSymbols: boolean,
+  useBuiltIns: boolean,
+): Generator<Transformed, void, void> {
+  const stack: Item[] = [];
+  const rootRight = right;
+  // The stack holds patterns that we don't known whether they contain private key
+  stack.push({
+    left,
+    right,
+    restExcludingKeys: initRestExcludingKeys(left),
+  });
+  let item: Item;
+  while ((item = stack.pop()) !== undefined) {
+    const { restExcludingKeys } = item;
+    let { left, right } = item;
+    const searchPrivateKey = privateKeyPathIterator(left).next();
+    if (searchPrivateKey.done) {
+      if (restExcludingKeys?.length > 0) {
```

Comment:
Is this handling something like the following? Turning `{ first, #priv: priv, ...rest }` into `{ first }, priv = #priv, { …rest }` would break the `…rest`'s exclude keys. So wee need to handle the case where something has been excluded. But in the case of `{ #priv: priv, first, …rest }`, it doesn't matter because `priv = #priv, { first, …rest }` still has the correct behavior.

Reply:
Yes. The `growRestExcludingKeys` will exclude private keys:

https://github.com/babel/babel/pull/14304/files#diff-5742fd1087b16ca97d1e7d8fbf176f27ad52d323f5e281db9bc349bd257efd26R68

> But in the case of { #priv: priv, first, …rest }, it doesn't matter because priv = #priv, { first, …rest } still has the correct behavior.

Yes. The plugin should already handle this case, I will add new tests.